tepeszCodes/document-management-system
# Chore: Setup Knex.js

#### What does this PR do?
- sets up Knex.js, migration and seeding
- adds CRUD operations template class
- adds models for each table in database
- defines most of the routes defined in a [template document](https://docs.google.com/document/d/1cZqCmaFP--tI8nkjEOGY1x-Z-Wdn1oEUOsxeWACGlac/edit) and adds services handling conntections
#### Description of Task to be completed?
Setup Knex.js
#### How should this be manually tested?
To test if Knex.js is added properly and if migrations and seeding are set up, first run `npm run migrate:rollback`, this will drop all the tables in development db. Then, run `npm run migrate:latest` to add new tables and `npm run seed` to populate them. You can check if tables are created and populated correctly using RDBMS of your choice.

To test if routes, services and models are set up propetly, run `npm run dev` script to start up the server and use Postman to test if each endpoint defined in `documentsRoute.js` and `usersRoute.js` behaves as expected. Server expects any data to be send in a JSON format and responds with data in the same format.
#### Any background context you want to provide?
I may have overinterpreted what I was supposed to do in this task. I was following the [template document](https://docs.google.com/document/d/1cZqCmaFP--tI8nkjEOGY1x-Z-Wdn1oEUOsxeWACGlac/edit) attached in the *Data should be persisted to a RDBMS storage* Trello card, but after adding all the models, services and routes I realized it was probably a template for the entire project, not just this one task. But what's done is done, I don't want to roll back so many commits that I have spent a lot of time on, so hopefully the way I implemented my solutions doesn't go against your plans for the next tasks.

Also, I implemented the tables in db a little bit differently - I moved `username` and `password` from `user_logins` to `user_profiles` table, as it seems to me to fall logically in the same category - general info about the user - whereas data stored in `user_logins` should, I think, contain only information about user logins, not their details.
#### Questions:
- Have I followed the barrel pattern correctly with all the new files?
- Are there any security concerns with storing all user details - username, email, password and fullname - in the one table?
- How should modules with classes be exported - should I instantiate a new object at the bottom of the file and export it (that's the way I did it in this PR), or export just the class and create new object in the files I import it to instead (this seems to violate the DRY rule, but maybe there are other reasons to do it this way)?
